cloud_images/lib/el7/dcos_vol_setup.sh

```diff
@@ -63,15 +64,22 @@ function main {
   fi
 
   noncritical echo -n "Waiting for $device to come online"
-  until test -b "$device"; do noncritical sleep 1; noncritical echo -n .; done
+  retry=5
+  until test -b "$device"; do
+    noncritical sleep 1; noncritical echo -n .;
+    let retry=retry-1
+    if [ $retry -eq 0 ]; then
+     noncritical exit 0
+    fi
+  done
```

Comment:
How does this change support for existing environments. Currently if a device isn't found the setup is considered broken and the service will fail. In this scenario a missing Mesos work_dir might result in the node coming up. The Admin will think everything was set up correctly until the node begins filling up the root disk device, for instance. Also, how is this change related to m5 support? I'd think a behavior change like this should be it's own PR.

Reply:
Yes, this is necessary to support m5 instances. Reason is that we now have more potential distinct device names. My changes split the one 'Oneshot' unit into dedicated per-device units.

**Example:**  The resulting AMIs will boot on an m4 instance, with the non-root EBS devices showing up as `/dev/xvde` trough `/dev/xvdh`, resulting in the four units for `xvd*` devices to complete and the four units for `nvme*n1` to _fail_ and vice versa on an m5 instance type.
The five second timeout showed to reliably work over at least 25 instance creations I did while testing this.

Reply:
Where/how do we make sure then that e.g. /var/log and /var/lib/docker is actually a volume and not the root device? I'm thinking of https://jira.mesosphere.com/browse/DCOS_OSS-1043 where a user had a small number of nodes fail during volume setup in his 100 node cluster due to a timing bug that rarely triggered. Also once we have the final solution here we should make sure to test it on a couple of hundred node cluster (I usually spin up 200 node clusters for one-off tests like this).

Reply:
> Where/how do we make sure then that e.g. /var/log and /var/lib/docker is actually a volume and not the root device?

We don't.
But unfortunately I do not see a way to get the same behavior as before. The only way to support both previous and current generation EC2 instances with their different hardware and therefore device names is to check for the presence of any  device. Not finding the other types of devices is expected and not a failure. And yes, I looked at leveraging udev for device discovery and potentially symlinking and spent multiple days trying to get it to work reliably... to no avail.

The current approach in this PR is the most simplistic solution I could come up with to keep supporting multiple generations of EC2 types. Not having one unified AMI but multiple type-generation specific ones is still an option though.